### PR TITLE
Add `CloudState` to be passed to `Cloud#provision` and `Cloud#canProvision` methods.

### PR DESCRIPTION
**Pull request by Vlatombe**
This gives more insight to the Cloud implementation to ongoing
provisioning operations and helps implement better control on the number
of nodes provisioned when provisioning limits are defined.

* Downstream PR https://github.com/jenkinsci/kubernetes-plugin/pull/835

### Proposed changelog entries

* Cloud implementations are given more context about ongoing planned nodes.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/slaves/Cloud.java
Comment (jglick) [inline core/src/main/java/hudson/slaves/Cloud.java:262]:
```suggestion
     * @since TODO
```

File: core/src/main/java/hudson/slaves/Cloud.java
Comment (res0nance) [inline core/src/main/java/hudson/slaves/Cloud.java:293]: Curious as to why this has to be synchronized

Comment (Vlatombe) [reply core/src/main/java/hudson/slaves/Cloud.java:293]: It doesn't need to be, this comes from copy-paste from `StrategyState`. In `CloudState` it's just readonly so I'll remove it.

File: core/src/main/java/hudson/slaves/Cloud.java
Comment (Vlatombe) [inline core/src/main/java/hudson/slaves/Cloud.java:167]: Found a problem with this aproach if there are several levels of subclassing. I wonder if there is already a utility covering this case in the Jenkins codebase...

Comment (Vlatombe) [reply core/src/main/java/hudson/slaves/Cloud.java:167]: `Util.isOverridden`

**Review by rsandell — APPROVED** [review]

File: core/src/main/java/hudson/slaves/Cloud.java
Comment (MRamonLeon) [inline core/src/main/java/hudson/slaves/Cloud.java:171]: You can create a private method to avoid duplicating this method, kinda `callIfOverriden` passing the code to call. WDYT?

**Review by MRamonLeon — COMMENTED** [review]
A comment to avoid duplicating the code. Some tests calling the deprecated and new methods would add confidence. Not sure if it's too hard.

File: core/src/main/java/hudson/slaves/Cloud.java
Comment (Vlatombe) [reply core/src/main/java/hudson/slaves/Cloud.java:171]: Good idea.

**Review by MRamonLeon — APPROVED** [review]

**Review by varyvol — APPROVED** [review]

**Review by alecharp — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by oleg-nenashev** [comment]
Looks like it caused https://issues.jenkins-ci.org/browse/JENKINS-63828 . CC @Vlatombe

File: core/src/main/java/hudson/slaves/Cloud.java
Comment (jglick) [inline core/src/main/java/hudson/slaves/Cloud.java:206]: Yet this is still called by https://github.com/jenkinsci/jenkins/blob/d99855db02eea957cb3bf8fad3af8b24638d2007/core/src/main/java/hudson/model/Label.java#L251-L266 and it is not clear that replacing that call with the new overload passing in a dummy `additionalPlannedCapacity` (say, of 1) would be correct. `Label.clouds` is cached for several minutes until https://github.com/jenkinsci/jenkins/blob/d99855db02eea957cb3bf8fad3af8b24638d2007/core/src/main/java/jenkins/model/Jenkins.java#L1006-L1011 and is used in determining `isEmpty` and thus `Jenkins.getLabels()`, which ought to be as stable as possible.

The Javadoc is unclear about whether the question is about whether the cloud _could_ provision such a label in general, because of its static configuration, or whether it _can_ provision that label right now (with a specific number of executors) based on live status information. So far as I can tell, common `Cloud` impls interpret it as the former (static configuration), but this is thrown into question by the new overload.

* https://github.com/jenkinsci/ec2-plugin/blob/eb62bd2d9548c5a64d425b6f25528c97a26f9a4e/src/main/java/hudson/plugins/ec2/EC2Cloud.java#L1171-L1174
* https://github.com/jenkinsci/kubernetes-plugin/blob/a2c22102c6c0e3eaa98361573868b3063de75476/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java#L751-L754
* https://github.com/jenkinsci/google-compute-engine-plugin/blob/9020cf2b745383e53c56274c652b55ee70c1df93/src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloud.java#L372-L380

Perhaps we should undeprecate `canProvision(Label)`, clarifying in its documentation that it should solely take into account static configuration (essentially whether the label is even mentioned in cloud configuration), and clarifying in documentation for `canProvision(CloudState)` that it _may_ (but need not) take into account dynamic runtime information?

File: core/src/main/java/hudson/slaves/CloudProvisioningListener.java
Comment (daniel-beck) [inline core/src/main/java/hudson/slaves/CloudProvisioningListener.java:63]: Missing `@since TODO`

**Pull request merged**
